### PR TITLE
feat(lockscript：add verify and test)

### PR DESCRIPTION
**Pull request by fpChan**

### CODE REVIEW

File: contracts/toCKB-lockscript/src/main.rs [hunk 1]
```diff
@@ -33,40 +30,49 @@ fn entry() -> i8 {
...
-        match err {
-            IndexOutOfBound => Self::IndexOutOfBound,
-            ItemMissing => Self::ItemMissing,
-            LengthNotEnough(_) => Self::LengthNotEnough,
-            Encoding => Self::Encoding,
-            Unknown(err_code) => panic!("unexpected sys error {}", err_code),
+fn verify() -> Result<(), Error> {
+    // load current lock_script hash
+    let script_hash = load_script_hash().unwrap();
+    let args: Bytes = load_script().unwrap().args().unpack();
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:40]:
```suggestion
    let script_hash = load_script_hash()?;
    let args: Bytes = load_script()?.args().unpack();
```

File: contracts/toCKB-lockscript/src/main.rs [hunk 2]
```diff
@@ -33,40 +30,49 @@ fn entry() -> i8 {
...
+    let cell_source = vec![Source::Input, Source::Output];
+
+    for &source in cell_source.iter() {
+        for i in 0.. {
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:44]: should refactor

File: tests/src/toCKB_lockscript/mod.rs [hunk 3]
```diff
@@ -1 +1,4 @@
+#[cfg(test)]
+mod tests;
 
+pub mod utils;
```
Comment (mkxbl) [inline tests/src/toCKB_lockscript/mod.rs:4]: no need

File: contracts/toCKB-lockscript/src/main.rs [hunk 4]
```diff
@@ -5,22 +5,19 @@
 #![feature(panic_info_message)]
 #![allow(non_snake_case)]
 
-// Import from `core` instead of from `std` since we are in no-std mode
-use core::result::Result;
-
-// Import heap related library from `alloc`
-// https://doc.rust-lang.org/alloc/index.html
-use alloc::{vec, vec::Vec};
+mod utils;
```
[the remark this answers is not in the document]
Comment (fpChan) [reply contracts/toCKB-lockscript/src/main.rs]: okay， done

File: tests/src/toCKB_lockscript/mod.rs (hunk 3 above)
Comment (fpChan) [reply tests/src/toCKB_lockscript/mod.rs]: done

File: contracts/toCKB-lockscript/src/main.rs [hunk 5]
```diff
@@ -57,16 +51,36 @@ impl From<SysError> for Error {
...
-    let tx_hash = load_tx_hash()?;
-    debug!("tx hash is {:?}", tx_hash);
+    verify_source_cells(Source::Input, script_hash, args.clone())?;
+    verify_source_cells(Source::Output, script_hash, args)?;
 
-    let _buf: Vec<_> = vec![0u8; 32];
+    Ok(())
+}
+
+fn verify_source_cells(source: Source, script_hash: [u8; 32], args: Bytes) -> Result<(), Error> {
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:67]:
```suggestion
fn verify_toCKB_cells(source: Source, script_hash: &[u8; 32], args: &[u8]) -> Result<(), Error> {
```

File: contracts/toCKB-lockscript/Cargo.toml [hunk 6]
```diff
@@ -7,6 +7,7 @@ edition = "2018"
 
 [dependencies]
 ckb-std = "0.6.0"
+hex = { version = "0.4", default-features = false }
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/Cargo.toml:10]:
```suggestion
```

File: contracts/toCKB-lockscript/src/main.rs [hunk 7]
```diff
@@ -57,16 +51,36 @@ impl From<SysError> for Error {
...
-    // remove below examples and write your code here
+    verify()
+}
 
-    let script = load_script()?;
-    let args: Bytes = script.args().unpack();
-    debug!("script args is {:?}", args);
+fn verify() -> Result<(), Error> {
+    let script_hash = load_script_hash()?;
+    let args: Bytes = load_script()?.args().unpack();
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:59]:
```suggestion
    let args = load_script()?.args().unpack().as_ref();
```

File: contracts/toCKB-lockscript/src/main.rs [hunk 8]
```diff
@@ -57,16 +51,36 @@ impl From<SysError> for Error {
...
 
-    let tx_hash = load_tx_hash()?;
-    debug!("tx hash is {:?}", tx_hash);
+    verify_source_cells(Source::Input, script_hash, args.clone())?;
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:61]:
```suggestion
    verify_toCKB_cells(Source::Input, &script_hash, args)?;
```

File: contracts/toCKB-lockscript/src/main.rs [hunk 9]
```diff
@@ -57,16 +51,36 @@ impl From<SysError> for Error {
...
+    verify_source_cells(Source::Output, script_hash, args)?;
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:62]:
```suggestion
    verify_toCKB_cells(Source::Output, &script_hash, args)?;
```

File: tests/src/toCKB_lockscript/mod.rs [hunk 10]
```diff
@@ -1 +1,192 @@
...
+        always_success_out_point,
+    );
+}
+
+fn build_output_cell(cell_case: CellCase) -> CellOutput {
+    let (mut context, toCKB_lockscript_out_point, always_success_out_point) =
+        load_context_and_out_points();
+
+    // prepare scripts
+    let fake_script = context
```
Comment (mkxbl) [inline tests/src/toCKB_lockscript/mod.rs:58]:
```suggestion
    let mock_toCKB_typescript = context
```

File: tests/src/toCKB_lockscript/mod.rs [hunk 11]
```diff
@@ -1 +1,192 @@
...
+    let always_success_out_point = context.deploy_cell(ALWAYS_SUCCESS.clone());
+
+    return (
+        context,
+        toCKB_lockscript_out_point,
+        always_success_out_point,
+    );
+}
+
+fn build_output_cell(cell_case: CellCase) -> CellOutput {
```
Comment (mkxbl) [inline tests/src/toCKB_lockscript/mod.rs:53]:
```suggestion
fn build_cell(cell_case: CellCase) -> CellOutput {
```

File: tests/src/toCKB_lockscript/mod.rs [hunk 12]
```diff
@@ -1 +1,192 @@
...
+pub enum Error {
+    IndexOutOfBound = 1,
+    ItemMissing,
+    LengthNotEnough,
+    Encoding,
+    // Add customized errors here...
+    InvalidToCKBCell,
+}
+
+fn get_lock_script_args(type_script: Script, is_ToCKBCell: bool) -> Bytes {
```
Comment (mkxbl) [inline tests/src/toCKB_lockscript/mod.rs:30]: no need

Comment (fpChan) [reply tests/src/toCKB_lockscript/mod.rs:30]: It's used for a  test case which  there is no toCKB cell  which  `lockscript args`  are not equal to the `typescript hash` 。

File: contracts/toCKB-lockscript/src/main.rs [hunk 13]
```diff
@@ -5,22 +5,15 @@
...
-// Import CKB syscalls and structures
-// https://nervosnetwork.github.io/ckb-std/riscv64imac-unknown-none-elf/doc/ckb_std/index.html
+use ckb_std::high_level::{load_cell_type, load_script};
 use ckb_std::{
+    ckb_constants::Source,
     ckb_types::{bytes::Bytes, prelude::*},
     debug, default_alloc, entry,
     error::SysError,
-    high_level::{load_script, load_tx_hash},
+    high_level::{load_cell_lock_hash, load_script_hash},
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:14]:
```suggestion
    high_level::{load_cell_lock_hash, load_script_hash, load_cell_type, load_script},
```

File: contracts/toCKB-lockscript/src/main.rs [hunk 14]
```diff
@@ -57,16 +51,41 @@ impl From<SysError> for Error {
...
-    let args: Bytes = script.args().unpack();
-    debug!("script args is {:?}", args);
+fn verify() -> Result<(), Error> {
+    let script_hash = load_script_hash()?;
+    let args: Bytes = load_script()?.args().unpack();
 
-    let tx_hash = load_tx_hash()?;
-    debug!("tx hash is {:?}", tx_hash);
+    verify_toCKB_cells(Source::Input, &script_hash, args.as_ref())?;
+    verify_toCKB_cells(Source::Output, &script_hash, args.as_ref())?;
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:62]: try another way: use Source::Group to find cells, and verify all cell_typescript_hash[0..54] == args

Comment (fpChan) [reply contracts/toCKB-lockscript/src/main.rs]: done, just use the GroupInput

File: contracts/toCKB-lockscript/src/main.rs [hunk 15]
```diff
@@ -57,16 +50,15 @@ impl From<SysError> for Error {
...
-    let tx_hash = load_tx_hash()?;
-    debug!("tx hash is {:?}", tx_hash);
+    verify()
+}
 
-    let _buf: Vec<_> = vec![0u8; 32];
+fn verify() -> Result<(), Error> {
+    let args: Bytes = load_script()?.args().unpack();
+    let count = QueryIter::new(load_cell_type, Source::GroupInput)
+        .filter(|type_script_opt| (type_script_opt.clone().unwrap().as_slice()[0..54] != args[..]))
```
Comment (mkxbl) [inline contracts/toCKB-lockscript/src/main.rs:59]: 1. 
```suggestion
        .filter(|type_script_opt| type_script_opt.is_some())
        .filter(|type_script_opt| (type_script_opt.clone().unwrap().as_slice()[0..54] != args[..]))
```

2. the logic is wrong
should verify count == 0

Comment (fpChan) [reply contracts/toCKB-lockscript/src/main.rs]: done

**Pull request merged**